File: stack/js/decompressBraces.js
const decompressBraces = (s) => {
  const nums = '123456789';
  const stack = [];
  for (let char of s) {
    if (nums.includes(char)) {
      stack.push(Number(char));
    } else {
      if (char === '}') {
        let segment = '';
        while (typeof stack[stack.length - 1] !== 'number') {
          const popped = stack.pop();
          segment = popped + segment;
        }
        const number = stack.pop();
        stack.push(repeat(segment, number));
      } else if (char !== '{') {
        stack.push(char);
      }
    }
  }
  
  return stack.join('');
};

const repeat = (str, n) => {
  let result = '';
  for (let i = 0; i < n; i += 1) {
    result += str;
  }
  return result;
};

File: stack/java/DecompressBraces.java
import java.util.ArrayDeque;
import java.util.Deque;

public final class DecompressBraces {
  private DecompressBraces() {
  }

  public static String run(String input) {
    Deque<String> stack = new ArrayDeque<>();
    for (char ch : input.toCharArray()) {
      if (Character.isDigit(ch)) {
        stack.push(String.valueOf(ch));
      } else if (ch == '{') {
        continue;
      } else if (ch == '}') {
        StringBuilder segment = new StringBuilder();
        while (!stack.isEmpty() && !isNumber(stack.peek())) {
          segment.insert(0, stack.pop());
        }
        int count = Integer.parseInt(stack.pop());
        stack.push(segment.toString().repeat(count));
      } else {
        stack.push(String.valueOf(ch));
      }
    }
    StringBuilder result = new StringBuilder();
    for (String part : stack.descendingIterator()) {
      result.append(part);
    }
    return result.toString();
  }

  private static boolean isNumber(String value) {
    return value.length() == 1 && Character.isDigit(value.charAt(0));
  }
}
